### PR TITLE
bpo-34100: Merge constants recursively

### PR DESCRIPTION
**Pull request by methane**
There are some same consts in a module.  This commit merges them into
single instance.  It reduces number of objects in memory after loading modules.

https://bugs.python.org/issue34100

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2018-07-27-20-04-52.bpo-34100.ypJQX1.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Compiler now merges constans in tuples and frozensets recursively. Code
```
Comment (asottile) [inline Misc/NEWS.d/next/Core and Builtins/2018-07-27-20-04-52.bpo-34100.ypJQX1.rst]: minor typo `constans` => `constants`

**Review by asottile — COMMENTED** [review]
didn't really review the code -- sorry about the drive-by!

File: Python/compile.c [hunk 2]
```diff
@@ -1179,18 +1188,121 @@ compiler_add_o(struct compiler *c, PyObject *dict, PyObject *o)
     return arg;
 }
 
+// Merge const *o* recursively and return constant key object.
+static PyObject*
+merge_consts(struct compiler *c, PyObject *o)
```
Comment (vstinner) [inline Python/compile.c]: I suggest to add something to the function name to make it more explicitly that this one is recursive, like:

merge_consts_recursive()

To distinguish it from merge_const_tuple() which is not recursive.

Note: const or consts?

File: Python/compile.c [hunk 3]
```diff
@@ -161,6 +161,7 @@ struct compiler {
     int c_interactive;           /* true if in interactive mode */
     int c_nestlevel;
 
+    PyObject *c_const_cache;     /* all constants, including names tuple */
```
Comment (vstinner) [inline Python/compile.c]: Please document that it's a dict. Maybe "Memo dictionary used to merge all constants".

**Review by vstinner — COMMENTED** [review]
I like the overall idea. But I don't have the bandwidth to review reference counting and other subtle details.

File: Python/compile.c [hunk 4]
```diff
@@ -285,9 +287,16 @@ compiler_init(struct compiler *c)
...
 
+    c->c_const_cache = PyDict_New();
+    if (!c->c_const_cache) {
+        return 0;
+    }
+
     c->c_stack = PyList_New(0);
-    if (!c->c_stack)
+    if (!c->c_stack) {
+        Py_DECREF(c->c_const_cache);
```
Comment (vstinner) [inline Python/compile.c]: Use maybe Py_CLEAR() to keep c consistent (otherwise, compiler_free() may crash).

File: Python/compile.c [hunk 5]
```diff
@@ -1179,18 +1189,121 @@ compiler_add_o(struct compiler *c, PyObject *dict, PyObject *o)
...
-    t = _PyCode_ConstantKey(o);
-    if (t == NULL)
+    PyObject *key = merge_consts_recursive(c, o);
+    if (key == NULL) {
         return -1;
+    }
 
-    arg = compiler_add_o(c, c->u->u_consts, t);
-    Py_DECREF(t);
+    Py_ssize_t arg = compiler_add_o(c, c->u->u_consts, key);
```
Comment (vstinner) [inline Python/compile.c:1305]: note: I like moving definitions of variables when they are used thanks to C99 ;-)

**Review by vstinner — APPROVED** [review]
LGTM.

cc @serhiy-storchaka

File: Python/compile.c [hunk 6]
```diff
@@ -5380,6 +5493,38 @@ compute_code_flags(struct compiler *c)
     return flags;
 }
 
+// Merge *tuple* with constant cache.
+// Unlike merge_consts_recursive(), this function doesn't work recursively.
+static int
+merge_const_tuple(struct compiler *c, PyObject **tuple)
+{
+    Py_ssize_t len = PyTuple_GET_SIZE(*tuple);
+    if (len == 0) {
```
Comment (vstinner) [inline Python/compile.c]: I'm not 100% sure that all empty tuples are the tuple singleton. Would it hurt to double check that we already have the singleton by removing this special case?

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by methane** [comment]
Thank you for review.

... Travis seems very unresponsible...

**Pull request merged**